### ISSUE TITLE
Prefer exact file name matches in results

### ISSUE THREAD

**Issue by hovissimo** [ctrlpvim/ctrlp.vim#318]
My project has many files found by 'job_test.rb'
![image](https://cloud.githubusercontent.com/assets/6434414/19902198/09bd6d26-a027-11e6-8d60-9b77d29b0151.png)

There is however one file `test/unit/job_test.rb` whose filename is an exact match for what I typed.
![image](https://cloud.githubusercontent.com/assets/6434414/19902226/23a84f30-a027-11e6-87fb-795767434c5d.png)

I have to preface the filename with unit to get this file and it's a minor nuisance.

It would be nice if CtrlP prioritized any files where the query is an exact match for the file name.

**Comment by hpurmann** [comment]
Yeah! It used to do this. This is broken for me as well. It seems that it tries to match the longest path instead of the exact filename.

Even if I type "folder/subfolder/filename/extension" (without the slashes) it gives me all the other results with fuzzy matched characters instead of the actual file.

**Comment by mattn** [comment]
Default matcher of CtrlP is fuzzy match. You can choice matcher. See `:help g:ctrlp_match_func`

**Comment by hpurmann** [comment]
I think I poorly phrased my comment. I _do want_ fuzzy matching. But the results used to be sorted differently. In the past, a short, exact matching name would precede a longer one.

Unfortunately, I don't know what made this change. [My config makes use of the silver searcher](https://github.com/hpurmann/dotfiles/blob/db57a7f6d8554001c52290005ba916bdcd8c7ecb/nvim/init.vim#L312-L322). @hovissimo are you using `ag` as well?

**Comment by mattn** [comment]
please try to set `let g:ctrlp_path_sort = 1`.

**Comment by Yggdroot** [comment]
Please try [LeaderF](https://github.com/Yggdroot/LeaderF), you won't get disappointed.

**Comment by Porges** [comment]
@mattn I can't see a setting called `path_sort` in ctrlp.vim?

**Comment by mattn** [comment]
@Porges see here https://github.com/ctrlpvim/ctrlp.vim/blob/master/autoload/ctrlp.vim#L2493

**Comment by hpurmann** [comment]
(@mattn on a sidenote: linking to specific files is a lot more useful if you always get the URL as a blob [like this](https://github.com/ctrlpvim/ctrlp.vim/blob/d9eb5aecab74db300bdf2c76308b2c02c469cf5e/autoload/ctrlp.vim#L2493) instead of linking to `master`. You can do so by pressing `y`)

**Comment by hovissimo** [comment]
@hpurmann I'm not using ag yet.

@mattn `let g:ctrlp_path_sort = 1` makes no difference:
![image](https://cloud.githubusercontent.com/assets/6434414/20692277/b20108a2-b58b-11e6-8ead-539d75ff8e7d.png)

**Comment by hpurmann** [comment]
I just switched to [`fzf`](https://github.com/junegunn/fzf.vim) because of that annoying issue. `fzf` is looking good so far.

**Comment by hovissimo** [comment]
Any hope for this issue?  I'm still not getting sorted results despite settings `let g:ctrlp_path_sort = 1`.

**Comment by mattn** [comment]
do you set g:ctrlp_path_nolim = 1 too? if set the option, sort is disabled.

**Comment by mattn** [comment]
And if you use matcher, also sort is disabled.

**Comment by hovissimo** [comment]
`g:ctrlp_path_nolim` and `g:ctrlp_match_func` are undefined

**Comment by MightyCreak** [comment]
Same here. I'm not using `ag`, not defining `g:ctrlp_path_nolim` nor `g:ctrlp_match_fun`, but the results are still not very good (see the issue I closed #356)

**Comment by mattn** [comment]
One more limit. If you treat over 10000 files, some files will be truncated.

**Comment by brandon1024** [comment]
For future readers that find themselves here, you can also consider using Vim's builtin `matchfuzzy()` function as a replacement for ctrlp's default match function. It was quite easy to do and the results have been quite a bit better. 

```
" configure an improved matching function that orders by closest match
let g:ctrlp_match_func = {
	\ 'match': 'CtrlpFuzzySearch',
\ }

function! CtrlpFuzzySearch(items, pat, limit, mmode, ispath, exc, regex)
	if empty(a:pat)
		return items[:a:limit]
	endif

	if a:regex
		return filter(a:items, {idx, val -> val =~ a:pat})
	endif

	return reverse(matchfuzzy(a:items, a:pat, {
		\ "limit": a:limit,
	\ }))
endfunction
```

This should be slightly faster too given that `matchfuzzy()` is written in C, but that's purely speculative. The only slight downside is that this solution doesn't highlight text in the match window, a feature that ctrlp supports out of the box. If you're crafty and find a solution, please let me know :wink:
